crates/bevy_asset/src/handle.rs

```diff
@@ -1,472 +1,383 @@
+use crate::{
+    meta::MetaTransform, Asset, AssetId, AssetIndexAllocator, AssetPath, InternalAssetId,
+    UntypedAssetId,
+};
+use bevy_ecs::prelude::*;
+use bevy_reflect::{Reflect, TypePath, Uuid};
+use crossbeam_channel::{Receiver, Sender};
 use std::{
-    cmp::Ordering,
-    fmt::Debug,
+    any::TypeId,
     hash::{Hash, Hasher},
-    marker::PhantomData,
+    sync::Arc,
 };
 
-use crate::{
-    path::{AssetPath, AssetPathId},
-    Asset, Assets,
-};
-use bevy_ecs::{component::Component, reflect::ReflectComponent};
-use bevy_reflect::{std_traits::ReflectDefault, Reflect, ReflectDeserialize, ReflectSerialize};
-use bevy_utils::Uuid;
-use crossbeam_channel::{Receiver, Sender};
-use serde::{Deserialize, Serialize};
-
-/// A unique, stable asset id.
-#[derive(
-    Debug, Clone, Copy, Eq, PartialEq, Hash, Ord, PartialOrd, Serialize, Deserialize, Reflect,
-)]
-#[reflect_value(Serialize, Deserialize, PartialEq, Hash)]
-pub enum HandleId {
-    /// A handle id of a loaded asset.
-    Id(Uuid, u64),
-
-    /// A handle id of a pending asset.
-    AssetPathId(AssetPathId),
+/// Provides [`Handle`] and [`UntypedHandle`] _for a specific asset type_.
+/// This should _only_ be used for one specific asset type.
+#[derive(Clone)]
+pub struct AssetHandleProvider {
+    pub(crate) allocator: Arc<AssetIndexAllocator>,
+    pub(crate) drop_sender: Sender<DropEvent>,
+    pub(crate) drop_receiver: Receiver<DropEvent>,
+    pub(crate) type_id: TypeId,
 }
 
-impl From<AssetPathId> for HandleId {
-    fn from(value: AssetPathId) -> Self {
-        HandleId::AssetPathId(value)
-    }
+pub(crate) struct DropEvent {
+    pub(crate) id: InternalAssetId,
+    pub(crate) asset_server_managed: bool,
 }
 
-impl<'a> From<AssetPath<'a>> for HandleId {
-    fn from(value: AssetPath<'a>) -> Self {
-        HandleId::AssetPathId(AssetPathId::from(value))
+impl AssetHandleProvider {
+    pub(crate) fn new(type_id: TypeId, allocator: Arc<AssetIndexAllocator>) -> Self {
+        let (drop_sender, drop_receiver) = crossbeam_channel::unbounded();
+        Self {
+            type_id,
+            allocator,
+            drop_sender,
+            drop_receiver,
+        }
     }
-}
 
-impl<'a, 'b> From<&'a AssetPath<'b>> for HandleId {
-    fn from(value: &'a AssetPath<'b>) -> Self {
-        HandleId::AssetPathId(AssetPathId::from(value))
+    /// Reserves a new strong [`UntypedHandle`] (with a new [`UntypedAssetId`]). The stored [`Asset`] [`TypeId`] in the
+    /// [`UntypedHandle`] will match the [`Asset`] [`TypeId`] assigned to this [`AssetHandleProvider`].
+    pub fn reserve_handle(&self) -> UntypedHandle {
```

Comment:
would it make sense to provide a typed version of this? Rust is probably capable of removing the type check on a typed version, while might have more difficulties with this (making this `#[inline(always)]` could render the point moot)

Reply:
I don't see how that could be true, given that AssetHandleProvider is inherently "untyped". There is no type information for the compiler to work with to optimize this.